### DISCUSSION TITLE
[Chapter 07] 카프카 프로듀서 로깅 중복코드를 AOP로

### DISCUSSION THREAD

**Discussion by JoisFe** [cowaine/spring-boot-for-msa#77] [Chapter 07. Spring AOP, Testing, and Auto Confiugration] [answered]
질문이라기 보다는 이번 챕터를 통해 작업 중인 프로젝트의 중복 코드들을 AOP를 이용해 리팩토링한 내용을 공유하고자 합니다.

## 이전 상황
- 진행해야되는 프로세스의 각 단계가 있는데 각 단계별 완료가 될 때마다 
카프카 프로듀서를 통해 로깅을 보냄
- 모든 단계과 완료 되면 완료가 되었다는 내용 또한 카프카 프로듀서를 통해 보냄
- 만약 중간에 에러가 난다면 에러 내용 + 중단 되었다는 내용을 로깅으로 보냄

## 문제점

``` java

try {
  비즈니스 로직 서비스  --> 각 서비스 메서드에 producer.send(완료되면 완료 내용)

  비즈니스 로직 서비스 --> 각 서비스 메서드에 producer.send(완료되면 완료 내용)
} catch(Exception e) {

   producer.send(에러내용)
   producer.send(중단)
}

```

카프카 컨슈머 통해 로깅을 보내는 코드가 비즈니스 로직 코드내에 들어가 있으며, 각 비즈니스 코드마다 컨슈머를 통해 로깅을 보내는 코드가 중복되고 있었음

이러한 중복된 코드들을 AOP를 통해 관점으로 분리하였고 상황에 맞는 에너테이션을 만들어 포인트 컷 명시하였음 
``` java
@After("@annotation(~~)")
@AfterThrowing("@annotation(~~)")
```

따라서 핵심 관심사가 아닌 아닌 계속 반복되는 공통 비기능적 요구 사항인 로깅을 분리하여 중복코드를 제거할 수 있었고
이후 유지보수도 더 편리하게 되었습니다. 
코드도 깔끔해지니 핵심 로직을 보기도 편리했구요

구체적인 코드를 공유드릴 순 없지만
스터디를 통해 학습한 내용을 개인 프로젝트 혹은 사내에서 활용할 수 있으면 좋은 것 같아 이번 글을 통해 공유하고 싶었습니다.

조금 더 좋은 방향이 보인다던지, 비슷한 경험을 공유해주시면 더욱 감사할 것 같습니다.

**Comment by corock** [comment, marked as answer]
코드 스니펫 보면 전부 컨슈머로 로깅을 처리하는 것처럼 보이는데
아직 안 배우긴 했지만 스프링 이벤트로 보내서 리스너로 처리하는 방법도 있을 것 같습니다.

**Comment by Irisation23** [comment]
### 금주 CheckPoint

-  코드에 나온 내용 컨슈머를 프로듀서로 바꿔야 함.
